design/operater-gateway-support-migration-design.md

```diff
@@ -0,0 +1,88 @@
+- Includes GatewayClass controller
+  - Expects parameter ref field to be unset, otherwise GatewayClass is not admitted
+  - Admits GatewayClass that has the configured controller name
+- Includes Gateway controller
+  - Checks if Gateway references Contour owned GatewayClass
+  - Passes off reconciliation to dag package Gateway API processor
+  - Gateway is validated, status set on it and relevant *Routes
+
+### Current Problems
+- Contour requires Gateway namespace/name *and* controller name, the latter of which the operator does not provide
+- Contour expects GatewayClass parameters ref to be unset while Operator expects it to reference a Contour CRD
+- If the above is rectified, Contour and the Operator still will conflict on setting status
+- CI failures that occur because of conflicts between Contour and the Operator are not detected unless we manually look at the scheduled Operator CI job
+
+## Goals
+- Improve implementation in Contour and Operator to mitigate current issues
+- Improve development processes and practices to ensure this conflicting state does not occur again
+
+## Non Goals
+- Drop support for features or projects
+
+## Remediation Steps
+
+- Set up alerting on Operator CI failures
+  - So we catch issues sooner.
+- Operator sets controller name in Contour ConfigMap
+  - To ensure Contour does not immediately exit on startup.
+- Operator stops setting any status on Gateway API resources
+  - To ensure Contour and Operator do not "fight" over status
+- Contour does not reject GatewayClasses if parameters ref is set
+  - At this point, we should have a workable intermediate solution for the moment and can cut a patch release
+  - Reconciliation will be duplicated but this should "work" for now
+  - User creates GatewayClass/Contour CRD with GatewayClass name
+  - Operator reconciles, does not set status
+  - User creates Gateway referencing GatewayClass
+  - Operator reconciles, creates a Contour
+  - Contour reconciles GatewayClass and Gateway, sets status, etc.
+- Operator deprecates Contour CRD GatewayClassRef field
+  - Currently the gateway reference has to be a GatewayClass name which is then validated to have the correct (hardcoded) controller string
+  - We should instead ensure the CRD has a field for setting the controller string the instance of Contour will watch for and reconcile GatewayClasses for
+  - The Operator should also set status to invalid on Contour objects that try to use a GatewayClass controller string that already exists, as it doesn't really make sense for multiple instances of Contour to watch for the same Controller string
+- Contour no longer requires Gateway namespace/name to be configured
+  - This is a planned deprecation anyway
+  - Will help with next step
+- Remove GatewayClass and Gateway controllers from Operator
+  - Planned deprecation anyways
+  - Users will no longer create a GatewayClass that references a Contour CRD to get a Contour that is configured for Gateway API.
```

Comment:
a GatewayClass finds a Contour instance (kind:Contour) through its spec.parametersRef right?  But that was never required I thought, as in the Operator will create some default configured Contour for you if left unspecified.

Reply:
1. Create Contour CR
2. Operator Creates Contour instance
3. Newly created Contour instance looks at its config and spins up Envoy infra
4. Create GatewayClass/Gateway
5. Process resources into Envoy

Reply:
step 1 is manual by user right? and so is step2, everything onward is automatically handled by the deployed Operator

Reply:
> and so is step2

there should be a step between 1-2: User creates Gateway

step 2 is not manual, the Operator performs steps 2-5

> Operator will create some default configured Contour for you if left unspecified

the Operator never implemented this, it actually always required a parametersRef